utils_test.go

```diff
@@ -42,3 +50,202 @@ func TestValidStoragePathFormat(t *testing.T) {
+			assert.NilError(t, err)
+			path = tmpDir
+		}
+		return system.Lstat(path)
+	})
+
+	rootlessRuntimeDir, err = getRootlessRuntimeDir(rootlessUID)
+	assert.NilError(t, err)
+	assert.Assert(t, rootlessRuntimeDir == systemdRuntimeDir, "Rootless runtime dir should follow systemd dir.")
+
+	lstatGuard = monkey.Patch(system.Lstat, func(path string) (*system.StatT, error) {
+		lstatGuard.Unpatch()
+		defer lstatGuard.Restore()
+		if path == systemdRuntimeDir {
+			err = os.Chmod(tmpDir, 0777)
+			assert.NilError(t, err)
+			path = tmpDir
+		}
+		return system.Lstat(path)
+	})
+
+	rootlessRuntimeDir, err = getRootlessRuntimeDir(rootlessUID)
+	assert.NilError(t, err)
+	assert.Assert(t, rootlessRuntimeDir != systemdRuntimeDir, "Rootless runtime dir shouldn't follow invalid systemd dir.")
+
+	// testing systemd dir for init based os.
+
+	readFileGuard = monkey.Patch(ioutil.ReadFile, func(path string) ([]byte, error) {
+		readFileGuard.Unpatch()
+		defer readFileGuard.Restore()
+		if path == "/proc/1/comm" {
+			return []byte("init"), nil
+		}
+		return ioutil.ReadFile(path)
+	})
+
+	lstatGuard = monkey.Patch(system.Lstat, func(path string) (*system.StatT, error) {
+		lstatGuard.Unpatch()
+		defer lstatGuard.Restore()
+		if path == systemdRuntimeDir {
+			err = os.Chmod(tmpDir, 0700)
+			assert.NilError(t, err)
+			err = os.Chown(tmpDir, os.Getuid(), os.Getgid())
+			assert.NilError(t, err)
+			path = tmpDir
+		}
+		return system.Lstat(path)
+	})
+
+	rootlessRuntimeDir, err = getRootlessRuntimeDir(rootlessUID)
+	assert.NilError(t, err)
+	assert.Assert(t, rootlessRuntimeDir != systemdRuntimeDir, "Rootless runtime dir shouldn't follow systemd dir.")
+
+	// testing target temp dir.
+
+	targetTmpDir := fmt.Sprintf("%s/containers-user-%d", os.TempDir(), rootlessUID)
+
+	lstatGuard = monkey.Patch(system.Lstat, func(path string) (*system.StatT, error) {
+		lstatGuard.Unpatch()
+		defer lstatGuard.Restore()
+		if path == targetTmpDir {
+			err = os.Chmod(tmpDir, 0700)
+			assert.NilError(t, err)
+			err = os.Chown(tmpDir, os.Getuid(), os.Getgid())
+			assert.NilError(t, err)
+			path = tmpDir
+		}
+		return system.Lstat(path)
+	})
+
+	rootlessRuntimeDir, err = getRootlessRuntimeDir(rootlessUID)
+	assert.NilError(t, err)
+	assert.Assert(t, rootlessRuntimeDir == targetTmpDir, "Rootless runtime dir should follow target temp dir.")
+
+	lstatGuard = monkey.Patch(system.Lstat, func(path string) (*system.StatT, error) {
+		lstatGuard.Unpatch()
+		defer lstatGuard.Restore()
+		if path == targetTmpDir {
+			err = os.Chmod(tmpDir, 0777)
+			assert.NilError(t, err)
+			path = tmpDir
+		}
+		return system.Lstat(path)
+	})
+
+	rootlessRuntimeDir, err = getRootlessRuntimeDir(rootlessUID)
+	assert.NilError(t, err)
+	assert.Assert(t, rootlessRuntimeDir != targetTmpDir, "Rootless runtime dir shouldn't follow invalid target temp dir.")
+
+	// testing target home dir.
+
+	targetHomeDir := tmpDir + "/rundir"
+
+	monkey.Patch(homedir.Get, func() string {
```

Comment:
In this case, os.Setenv("HOME")

Reply:
I can't agree, this won't be cross platform. We shouldn't try to mock using any internal business logic of dependant software.

Reply:
Fine, then parametrize this path by splitting the implementation into two, as discussed for reading `/proc`

Reply:
Sure, that will be better solution.